documentation/project-docs/cli-guidelines.md

```diff
@@ -0,0 +1,216 @@
+If the command is run in a directory that contains multiple project or solution files, the command should fail with an error message
+indicating that the user should specify which file to use.
+
+Consider the use of an option like `--project <path to project or solution>` to allow for executing the command against a specific
+project or solution file at an arbitrary location - but be aware that many mechanisms in .NET are hierarchical from Current Working
+Directory and may have different behaviors when used in this detached mode.
+
+### Determining interactive sessions
+
+Commands should be able to determine if they are running in an interactive session or not. This is important for determining
+
+* whether to prompt the user for input or not,
+* what the default value for NuGet's `--interactive` option should be
+* how to format output (along with terminfo probing)
+
+In general the easiest way to check this is if the `stdin` file descriptor is a TTY, or in .NET Terms: `Console.IsInputRedirected` is `false`.
+
+## Interaction patterns
+
+### Support structured (JSON) output
+
+All commands should support structured (JSON) output. This is a key part of the CLI experience and allows for easier scripting and automation of the CLI.
+See the [`--format`](#output-modesformatting) option for more details. When structured output is requested, the command should output the structured
+data to `stdout` and any non-structured data to `stderr` - this is to make it easier to parse the structured data without having to deal with the
+non-structured data.
+
+### StdOut/StdErr usage
+
+StdErr is a perfectly reasonable output channel. It should be the default channel for
+
+* warnings and errors
+* verbose logging
+* outputs that aren't directly related to the command's primary purpose - for example 'workload manifest update' notices
+
+There is some contention here because older versions of PowerShell treat any stderr output as an error, but this is not the
+case in modern PowerShell. If you're writing a command that is intended to be used in PowerShell, you should be aware of this
+ and ensure that your command behaves correctly in both old and new versions of PowerShell - consider using Powershells various
+ [output streams][pwsh-output-streams] to encode the semantics of your output.
+
+### Advanced terminal output capabilities
+
+You are encouraged to make use of advanced terminal capabilities like colors, weights, and other formatting options to make your output
+more readable. Instead of writing entire lines of text as a single color, consider how the use of color and formatting present the
+output to the user. When performing long-running operations, provide a progress bar or ticker or some other form of feedback to the user.
+
+DO NOT rely entirely on color to convey information. Colorblind users will not be able to distinguish between different colors. Instead,
+consider color, symbols, and layout holistically to convey information.
```

Comment:
I also like this suggestion. Is there any code to show examples of using this? Might be useful to have linked examples (code files in the repo) for each of these patterns.